%%
%% This is file `sample-manuscript.tex',
%% generated with the docstrip utility.
%%
%% The original source files were:
%%
%% samples.dtx  (with options: `manuscript')
%% 
%% IMPORTANT NOTICE:
%% 
%% For the copyright see the source file.
%% 
%% Any modified versions of this file must be renamed
%% with new filenames distinct from sample-manuscript.tex.
%% 
%% For distribution of the original source see the terms
%% for copying and modification in the file samples.dtx.
%% 
%% This generated file may be distributed as long as the
%% original source files, as listed above, are part of the
%% same distribution. (The sources need not necessarily be
%% in the same archive or directory.)
%%
%% Commands for TeXCount
%TC:macro \cite [option:text,text]
%TC:macro \citep [option:text,text]
%TC:macro \citet [option:text,text]
%TC:envir table 0 1
%TC:envir table* 0 1
%TC:envir tabular [ignore] word
%TC:envir displaymath 0 word
%TC:envir math 0 word
%TC:envir comment 0 0
%%
%%
%% The first command in your LaTeX source must be the \documentclass command.
%%%% Small single column format, used for CIE, CSUR, DTRAP, JACM, JDIQ, JEA, JERIC, JETC, PACMCGIT, TAAS, TACCESS, TACO, TALG, TALLIP (formerly TALIP), TCPS, TDSCI, TEAC, TECS, TELO, THRI, TIIS, TIOT, TISSEC, TIST, TKDD, TMIS, TOCE, TOCHI, TOCL, TOCS, TOCT, TODAES, TODS, TOIS, TOIT, TOMACS, TOMM (formerly TOMCCAP), TOMPECS, TOMS, TOPC, TOPLAS, TOPS, TOS, TOSEM, TOSN, TQC, TRETS, TSAS, TSC, TSLP, TWEB.
% \documentclass[acmsmall]{acmart}

%%%% Large single column format, used for IMWUT, JOCCH, PACMPL, POMACS, TAP, PACMHCI
% \documentclass[acmlarge,screen]{acmart}

%%%% Large double column format, used for TOG
% \documentclass[acmtog, authorversion]{acmart}

%%%% Generic manuscript mode, required for submission
%%%% and peer review
\documentclass[sigconf]{acmart}
%% Fonts used in the template cannot be substituted; margin 
%% adjustments are not allowed.
%%
%% \BibTeX command to typeset BibTeX logo in the docs
\AtBeginDocument{%
  \providecommand\BibTeX{{%
    \normalfont B\kern-0.5em{\scshape i\kern-0.25em b}\kern-0.8em\TeX}}}

\input{packages}
\input{acronyms}

%%%%%%%%%%%%%%%
% Custom floats
%%%%%%%%%%%%%%%

%\newfloat{algorithm}{thp}{lop}
%l\floatname{algorithm}{Algorithm}

%% Rights management information.  This information is sent to you
%% when you complete the rights form.  These commands have SAMPLE
%% values in them; it is your responsibility as an author to replace
%% the commands and values with those provided to you when you
%% complete the rights form.

\copyrightyear{2023}
\acmYear{2023}
\setcopyright{cc}
\setcctype{by-sa}
\acmConference[HEART 2023]{The International Symposium on Highly Efficient Accelerators and Reconfigurable Technologies 2023}{June 14--16, 2023}{Kusatsu, Japan}
\acmBooktitle{The International Symposium on Highly Efficient Accelerators and Reconfigurable Technologies 2023 (HEART 2023), June 14--16, 2023, Kusatsu, Japan}
\acmDOI{10.1145/3597031.3597050}
\acmISBN{979-8-4007-0043-9/23/06}

%%
%% For managing citations, it is recommended to use bibliography
%% files in BibTeX format.
%%
%% You can then either use BibTeX with the ACM-Reference-Format style,
%% or BibLaTeX with the acmnumeric or acmauthoryear sytles, that include
%% support for advanced citation of software artefact from the
%% biblatex-software package, also separately available on CTAN.
%%
%% Look at the sample-*-biblatex.tex files for templates showcasing
%% the biblatex styles.
%%

%%
%% The majority of ACM publications use numbered citations and
%% references.  The command \citestyle{authoryear} switches to the
%% "author year" style.
%%
%% If you are preparing content for an event
%% sponsored by ACM SIGGRAPH, you must use the "author year" style of
%% citations and references.
%% Uncommenting
%% the next command will enable that style.
%%\citestyle{authoryear}

%%
%% end of the preamble, start of the body of the document source.
\begin{document}

%%
%% The "title" command has an optional parameter,
%% allowing the author to define a "short title" to be used in page headers.
\title[]{Mutation Tree Reconstruction of Tumor Cells on FPGAs Using a Bit-Level Matrix Representation}

%%
%% The "author" command and its associated commands are used to define
%% the authors and their affiliations.
%% Of note is the shared affiliation of the first two authors, and the
%% "authornote" and "authornotemark" commands
%% used to denote shared contribution to the research.
\author{Jan-Oliver Opdenhövel}
\affiliation{
  \institution{Paderborn University, Paderborn Center for Parallel Computing}
  \city{Paderborn}
  \country{Germany}
}
\email{joo@mail.uni-paderborn.de}
\orcid{0000-0003-2314-2784}

\author{Christian Plessl}
\affiliation{
  \institution{Paderborn University, Paderborn Center for Parallel Computing}
  \city{Paderborn}
  \country{Germany}
}
\email{plessl@uni-paderborn.de}
\orcid{0000-0001-5728-9982}

\author{Tobias Kenter}
\affiliation{
  \institution{Paderborn University, Paderborn Center for Parallel Computing}
  \city{Paderborn}
  \country{Germany}
}
\email{kenter@uni-paderborn.de}
\orcid{0000-0002-5088-0267}

%%
%% By default, the full list of authors will be used in the page
%% headers. Often, this list is too long, and will overlap
%% other information printed in the page headers. This command allows
%% the author to define a more concise list
%% of authors' names for this purpose.
\renewcommand{\shortauthors}{Opdenhövel, Plessl, Kenter}

%%
%% The abstract is a short summary of the work to be presented in the
%% article.
\begin{abstract}
    Reconstructing the mutation history of a cancer cell in the form of a phylogenetic tree 
    from noisy genome sequencing data requires a likelihood model and a search or optimization strategy. The \ac{SCITE} software performs such a search via Monte Carlo Markov Chains that explore and evaluate different random update operations on a current candidate tree.
    In this work, we present an FPGA accelerator design (denoted as \ac{ffSCITE}) that implements all required operations with deterministic latency as bit-operations on a matrix representation. The largest design updates and counts 36100 set bits in each clock cycle and thus outperforms the original single-threaded software reference by up to 80$\times$.
\end{abstract}

%%
%% The code below is generated by the tool at http://dl.acm.org/ccs.cfm.
%% Please copy and paste the code instead of the example below.
%%
\begin{CCSXML}
<ccs2012>
   <concept>
       <concept_id>10010520.10010521.10010542.10010543</concept_id>
       <concept_desc>Computer systems organization~Reconfigurable computing</concept_desc>
       <concept_significance>500</concept_significance>
       </concept>
    <concept>
       <concept_id>10010405.10010444.10010450</concept_id>
       <concept_desc>Applied computing~Bioinformatics</concept_desc>
       <concept_significance>500</concept_significance>
       </concept>
   <concept>
       <concept_id>10010583.10010682.10010684</concept_id>
       <concept_desc>Hardware~High-level and register-transfer level synthesis</concept_desc>
       <concept_significance>100</concept_significance>
       </concept>
 </ccs2012>
\end{CCSXML}

\ccsdesc[500]{Computer systems organization~Reconfigurable computing}
\ccsdesc[500]{Applied computing~Bioinformatics}
\ccsdesc[100]{Hardware~High-level and register-transfer level synthesis}

%%
%% Keywords. The author(s) should pick words that accurately describe
%% the work being presented. Separate the keywords with commas.
\keywords{FPGA, Monte Carlo Markov Chain (MCMC), Bit-level parallelism, SYCL, oneAPI}

%%
%% This command processes the author and affiliation and title
%% information and builds the first part of the formatted document.
\maketitle
\acresetall

%\todo[inline]{What about open directions and related work? I have no open directions that appear interesting to me, and I have done little research on other stuff. Should I write something about that? And if so, how much? -> Yes, no paper completely without these two aspects. 
%Open directions can also serve as inspiration for other people, so the actual intention to do it yourself is not a necessity.
%Maybe there is no such closely related work that it needs its own section, but we can have a general discussion as part of the introduction. }

\input{sections/01-introduction}
\input{sections/02-background}
\input{sections/03-approach}
\input{sections/04-evaluation}
\input{sections/05-analysis}

\section{Related Work}

Typical FPGA accelerators for bioinformatics workloads~\cite{yu-kw-03a,zh-ta-07a,ja-la-08a} rely solely on highly data-parallel datapaths with customized bit-width. Similar to our work, Guo et al.~\cite{gu-la-19a} additionally consider an iterative step for a form of chain expansion around the data-parallel core. With this, they also encounter a latency challenge on the feedback of this step. Orthogonal to our solution of overlapping this latency with steps from other chains, they focus on minimizing the latency of the individual step by reordering the computation order.

On the algorithmic side, Kuipers et al.~\cite{ku-ja-17a} have extended the search space for possible mutation trees by softening the infinite-site assumption. This allows to reconstruct some mutation histories more accurately and could be considered in possible extensions of \ac{ffSCITE}.

\input{sections/07-conclusion}

\bibliographystyle{ACM-Reference-Format.bst}
\bibliography{literature} 

\end{document}
\endinput
%%
%% End of file `sample-authordraft.tex'.
